extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,361 @@
+			if (response && response.endPoints) {
+				let master = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+				this.addSqlMasterNode(master.endpoint, master.description);
+			}
+			return this.children as ControllerTreeNode[];
+		} catch (error) {
+			let e = error as IControllerError;
+			vscode.window.showErrorMessage(e.message);
+			return this.children as ControllerTreeNode[];
+		}
+	}
+
+	private static toIpAndPort(url: string): string {
+		if (!url) {
+			return;
+		}
+		return url.trim().replace(/ /g, '').replace(/^.+\:\/\//, '').replace(/:(\d+)$/, ',$1');
+	}
+
+	public addSqlMasterNode(endPointAddress: string, description: string): void {
+		let epFolder = this.getEndPointFolderNode();
+		epFolder.addChild(new SqlMasterNode({ endPointAddress, parent: epFolder, treeChangeHandler: this.treeChangeHandler, description }));
+	}
+
+	private getEndPointFolderNode(): FolderNode {
+		let label = 'SQL Servers';
+		let epFolderNode = this.children.find(e => e instanceof FolderNode && e.label === label);
+		if (!epFolderNode) {
+			epFolderNode = new FolderNode({ label, parent: this, treeChangeHandler: this.treeChangeHandler });
+			this.addChild(epFolderNode);
+		}
+		return epFolderNode as FolderNode;
+	}
+
+	public getTreeItem(): vscode.TreeItem {
+		let item: vscode.TreeItem = super.getTreeItem();
+		item.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+		return item;
+	}
+
+	public get url() {
+		return this._url;
+	}
+
+	public set url(url: string) {
+		this._url = url;
+	}
+
+	public get username() {
+		return this._username;
+	}
+
+	public set username(username: string) {
+		this._username = username;
+	}
+
+	public get password() {
+		return this._password;
+	}
+
+	public set password(pw: string) {
+		this._password = pw;
+	}
+
+	public get rememberPassword() {
+		return this._rememberPassword;
+	}
+
+	public set rememberPassword(rememberPassword: boolean) {
+		this._rememberPassword = rememberPassword;
+	}
+}
+
+export class FolderNode extends ControllerTreeNode {
+	constructor(properties?: {
+		label: string;
+		parent: ControllerTreeNode;
+		treeChangeHandler: IControllerTreeChangeHandler;
+	}) {
+		super(Object.assign({
+			description: properties.label,
+			nodeType: BdcItemType.folder,
+			iconPath: IconPath.folderNode
+		}, properties));
+	}
+}
+
+export class SqlMasterNode extends ControllerTreeNode {
+	private _role: string;
+	private _endPointAddress: string;
+	private _username: string;
+	private _password: string;
+
+	constructor(properties?: {
+		endPointAddress: string,
+		parent: ControllerTreeNode,
+		treeChangeHandler: IControllerTreeChangeHandler,
+		label?: string,
+		description?: string,
+	}) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.sqlMaster,
+			iconPath: IconPath.sqlMasterNode
+		}, properties));
+
+		this._endPointAddress = properties.endPointAddress;
+		this._username = 'sa';
+		this._role = 'sql-server-master';
```

Comment:
This username is always going to be sa? That can never change?

Reply:
Yes. That is the current requirement. Please mind that anything major like this is already covered by Kevin and Alan who watches really closely.

Reply:
@Charles - for now this is correct. This may change in the future but right now, it's a reasonable assumption. Worst case it's like Azure - we prompt for "default admin" and the user can fix.